File: app/notify.py
"""Alert creation plus outbound delivery channels."""

import logging

import httpx
from sqlalchemy.ext.asyncio import AsyncSession

from app.config import settings
from app.models import Alert

log = logging.getLogger("keyshort.notify")


async def send_telegram(text: str) -> None:
    if not settings.telegram_bot_token or not settings.telegram_chat_id:
        return
    url = f"https://api.telegram.org/bot{settings.telegram_bot_token}/sendMessage"
    try:
        async with httpx.AsyncClient(timeout=10) as client:
            await client.post(
                url,
                json={
                    "chat_id": settings.telegram_chat_id,
                    "text": text,
                    "parse_mode": "HTML",
                    "disable_web_page_preview": True,
                },
            )
    except Exception as exc:  # noqa: BLE001 - alerting must never break the request
        log.warning("telegram delivery failed: %s", exc)


async def notify(
    db: AsyncSession,
    *,
    user_id: int,
    api_key_id: int | None,
    kind: str,
    message: str,
    severity: str = "warning",
) -> None:
    db.add(
        Alert(
            user_id=user_id,
            api_key_id=api_key_id,
            kind=kind,
            severity=severity,
            message=message,
        )
    )
    await db.commit()
    prefix = "\u26a0\ufe0f" if severity == "warning" else "\U0001f6d1"
    await send_telegram(f"{prefix} <b>KeyShort</b>\n{message}")

File: src/notify.js
/**
 * Alert creation plus outbound delivery channels.
 */
const { settings } = require("./config");
const { Alert } = require("./models");

async function sendTelegram(text) {
  if (!settings.telegramBotToken || !settings.telegramChatId) return;
  const url = `https://api.telegram.org/bot${settings.telegramBotToken}/sendMessage`;
  try {
    const controller = new AbortController();
    const timeout = setTimeout(() => controller.abort(), 10000);
    try {
      await fetch(url, {
        method: "POST",
        headers: { "content-type": "application/json" },
        body: JSON.stringify({
          chat_id: settings.telegramChatId,
          text,
          parse_mode: "HTML",
          disable_web_page_preview: true,
        }),
        signal: controller.signal,
      });
    } finally {
      clearTimeout(timeout);
    }
  } catch (err) {
    // alerting must never break the request
    console.warn("telegram delivery failed:", err.message);
  }
}

async function notify({ userId, apiKeyId, kind, message, severity = "warning" }) {
  await Alert.create({
    userId,
    apiKeyId: apiKeyId ?? null,
    kind,
    severity,
    message,
  });
  const prefix = severity === "warning" ? "\u26a0\ufe0f" : "\ud83d\uded1";
  await sendTelegram(`${prefix} <b>KeyShort</b>\n${message}`);
}

module.exports = { sendTelegram, notify };
